File: packages/typescript/tests/observability.test.ts
import { describe, it, expect } from 'vitest';
import * as fs from 'fs';
import * as path from 'path';
import { parseSkp, buildScene } from '../src/index';
import { SkpParseError } from '../src/errors';
import { LogLevel } from '../src/observability';

/**
 * openskp exposes progress via optional onProgress/onLog callbacks (see
 * observability.ts) - silent by default, and raises SkpParseError with
 * structured location context (stage, recordIndex, tag, ...) on failure -
 * so a production pipeline can trace exactly where a model got stuck
 * instead of a bare stack trace.
 */
describe('Observability: progress callbacks + structured error context', () => {
  const filePath = path.join(__dirname, 'fixtures', 'capilla_quiroz_v17.skp');
  const buf = fs.readFileSync(filePath);
  const arrayBuffer = buf.buffer.slice(buf.byteOffset, buf.byteOffset + buf.byteLength) as ArrayBuffer;

  it('is silent by default (no callbacks fire without options)', () => {
    // Nothing to assert on directly - this just verifies parseSkp() works
    // fine with no options argument, matching the "silent by default"
    // design (no forced logging/printing of any kind).
    const model = parseSkp(arrayBuffer);
    expect(model.version).toBe('{17.0.18899}');
  });

  it('reports onLog messages for a real parse', () => {
    const messages: { level: LogLevel; message: string }[] = [];
    parseSkp(arrayBuffer, { onLog: (level, message) => messages.push({ level, message }) });

    expect(messages.some((m) => m.message.includes('Parsing legacy'))).toBe(true);
    expect(messages.some((m) => m.message.includes('Parse complete'))).toBe(true);
  });

  it('reports onLog messages for buildScene', () => {
    const messages: { level: LogLevel; message: string }[] = [];
    buildScene(arrayBuffer, { onLog: (level, message) => messages.push({ level, message }) });

    expect(messages.some((m) => m.message.includes('Building scene'))).toBe(true);
    expect(messages.some((m) => m.message.includes('Scene build complete'))).toBe(true);
  });

  it('raises SkpParseError with stage="zip_extract" for a corrupt file', () => {
    const bad = new Uint8Array(200).fill(0x41); // "AAAA..." - not a valid header
    const badBuffer = bad.buffer;

    expect(() => parseSkp(badBuffer)).toThrow(SkpParseError);
    try {
      parseSkp(badBuffer);
      expect.fail('expected parseSkp to throw');
    } catch (e) {
      expect(e).toBeInstanceOf(SkpParseError);
      expect((e as SkpParseError).stage).toBe('zip_extract');
    }
  });

  it('SkpParseError message includes structured context', () => {
    const err = new SkpParseError('boom', { stage: 'tlv_walk', recordIndex: 3, totalRecords: 10, tag: 'F601' });
    expect(err.message).toContain('stage=tlv_walk');
    expect(err.message).toContain('record=3/10');
    expect(err.message).toContain('tag=F601');
  });

  it('SkpParseError preserves the original error as .cause', () => {
    const original = new Error('inner failure');
    const wrapped = new SkpParseError('wrapped', { stage: 'tlv_walk', cause: original });
    expect(wrapped.cause).toBe(original);
  });
});

File: packages/cpp/tests/observability_test.cpp
#include <algorithm>
#include <filesystem>
#include <gtest/gtest.h>
#include <string>
#include <utility>
#include <vector>

#include <openskp/openskp.hpp>

#include "test_helpers.hpp"

using namespace openskp;

TEST(Errors, ContextAndCause) {
  auto cause = std::make_exception_ptr(std::logic_error("inner"));
  SkpParseError e("boom", ParseStage::tlv_walk, 3, 10, std::string("F601"), {}, {}, cause);
  EXPECT_NE(std::string(e.what()).find("stage=tlv_walk"), std::string::npos);
  EXPECT_NE(std::string(e.what()).find("record=3/10"), std::string::npos);
  EXPECT_EQ(e.cause(), cause);
}

TEST(Observability, CallbacksAreOptional) {
  EXPECT_THROW(SkpFile::from_buffer(ByteBuffer(20, 'A')).parse(), SkpParseError);
  EXPECT_NO_THROW(SkpFile::open(test::fixture("Untitled.skp")).parse());
}

TEST(Observability, ParseEmitsExpectedLogs) {
  std::vector<std::pair<LogLevel, std::string>> logs;
  ParseOptions options;
  options.log = [&](LogLevel level, std::string_view message) {
    logs.emplace_back(level, message);
  };

  SkpFile::open(test::fixture("capilla_quiroz_v17.skp")).parse(options);

  ASSERT_GE(logs.size(), 4u);
  EXPECT_EQ(logs.front().first, LogLevel::information);
  EXPECT_NE(logs.front().second.find("Parsing buffer"), std::string::npos);
  EXPECT_TRUE(std::any_of(logs.begin(), logs.end(), [](const auto& entry) {
    return entry.first == LogLevel::debug && entry.second.find("legacy MFC") != std::string::npos;
  }));
  EXPECT_TRUE(std::any_of(logs.begin(), logs.end(), [](const auto& entry) {
    return entry.second.find("Parse complete") != std::string::npos;
  }));
}

TEST(Observability, SceneBuildEmitsExpectedLogs) {
  std::vector<std::string> messages;
  ParseOptions options;
  options.log = [&](LogLevel, std::string_view message) { messages.emplace_back(message); };

  SkpFile::open(test::fixture("SU_File.skp")).build_scene(options);

  EXPECT_TRUE(std::any_of(messages.begin(), messages.end(), [](const auto& message) {
    return message.find("Building scene") != std::string::npos;
  }));
  EXPECT_TRUE(std::any_of(messages.begin(), messages.end(), [](const auto& message) {
    return message.find("Scene build complete") != std::string::npos;
  }));
}

TEST(Observability, ProgressIncludesLegacyDefinitions) {
  std::vector<ParseProgress> progress;
  ParseOptions options;
  options.progress = [&](const ParseProgress& value) { progress.push_back(value); };

  auto model = SkpFile::open(test::fixture("capilla_quiroz_v17.skp")).parse(options);

  ASSERT_FALSE(progress.empty());
  const auto found = std::find_if(progress.begin(), progress.end(), [](const auto& value) {
    return value.stage == ParseStage::legacy_defs;
  });
  ASSERT_NE(found, progress.end());
  EXPECT_EQ(found->current, model.definitions.size());
  EXPECT_EQ(found->total, model.definitions.size());
}

TEST(Errors, InvalidBufferReportsHeaderStage) {
  try {
    SkpFile::from_buffer(ByteBuffer(20, 'A')).parse();
    FAIL() << "Expected SkpParseError";
  } catch (const SkpParseError& error) {
    ASSERT_TRUE(error.stage().has_value());
    EXPECT_EQ(*error.stage(), ParseStage::header);
  }
}

TEST(ParserPaths, RejectsExistingFileWithWrongExtension) {
  const auto cmake_file = std::filesystem::path(OPENSKP_FIXTURES).parent_path() / "CMakeLists.txt";
  EXPECT_THROW(SkpFile::open(cmake_file), std::invalid_argument);
}
